File: app/src/main/java/com/vsp/internetspeedmeter/Room/UsageRepository.java
package com.vsp.internetspeedmeter.Room;

import android.content.Context;

import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import androidx.lifecycle.LiveData;

public class UsageRepository {

    /**
     * AsyncTask, which this used to spawn per write, is deprecated since API 30 and
     * its shared pool gives no ordering guarantee. A single serial executor keeps the
     * daily row's writes in order and off the main thread.
     */
    static final ExecutorService DB_EXECUTOR = Executors.newSingleThreadExecutor();

    private final UsageDao usageDao;
    private final LiveData<List<Usage>> allUsage;

    public UsageRepository(Context context) {
        UsageDatabase database = UsageDatabase.getInstance(context);
        usageDao = database.usageDao();
        allUsage = usageDao.getAllUsage();
    }

    /** Upsert: the date is the primary key and the DAO replaces on conflict. */
    public void insert(Usage usage) {
        DB_EXECUTOR.execute(() -> usageDao.insert(usage));
    }

    public void update(Usage usage) {
        DB_EXECUTOR.execute(() -> usageDao.update(usage));
    }

    public void delete(Usage usage) {
        DB_EXECUTOR.execute(() -> usageDao.delete(usage));
    }

    public void deleteAllNotes() {
        DB_EXECUTOR.execute(usageDao::deleteAll);
    }

    public LiveData<List<Usage>> getAllUsage() {
        return allUsage;
    }
}

File: app/src/main/java/com/vsp/internetspeedmeter/room/UsageRepository.kt
package com.vsp.internetspeedmeter.room

import android.content.Context
import androidx.lifecycle.LiveData
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

class UsageRepository(context: Context) {
    private val usageDao: UsageDao = UsageDatabase.getInstance(context).usageDao()
    val allUsage: LiveData<List<Usage>> = usageDao.getAllUsage()
    val allUsageFlow: Flow<List<Usage>> = usageDao.getAllUsageFlow()

    private val repositoryScope = CoroutineScope(
        SupervisorJob() + Dispatchers.IO.limitedParallelism(1)
    )
    private val writeMutex = Mutex()

    suspend fun upsert(usage: Usage) = withContext(Dispatchers.IO) {
        writeMutex.withLock {
            usageDao.upsert(usage)
        }
    }

    suspend fun insertAllIgnore(usages: List<Usage>) = withContext(Dispatchers.IO) {
        writeMutex.withLock {
            usageDao.insertAllIgnore(usages)
        }
    }

    suspend fun getUsageByDate(date: String): Usage? = withContext(Dispatchers.IO) {
        usageDao.getUsageByDate(date)
    }

    suspend fun getMonthlyMobileExcluding(monthYear: String, excludeDate: String): Long =
        withContext(Dispatchers.IO) {
            usageDao.getMonthlyMobileExcluding(monthYear, excludeDate)
        }

    suspend fun delete(usage: Usage) = withContext(Dispatchers.IO) {
        writeMutex.withLock {
            usageDao.delete(usage)
        }
    }

    suspend fun deleteAllNotes() = withContext(Dispatchers.IO) {
        writeMutex.withLock {
            usageDao.deleteAll()
        }
    }

    // Fire-and-forget helpers for background tasks
    fun insert(usage: Usage) {
        repositoryScope.launch {
            writeMutex.withLock {
                usageDao.upsert(usage)
            }
        }
    }

    fun update(usage: Usage) {
        repositoryScope.launch {
            writeMutex.withLock {
                usageDao.upsert(usage)
            }
        }
    }

    fun deleteAsync(usage: Usage) {
        repositoryScope.launch {
            writeMutex.withLock {
                usageDao.delete(usage)
            }
        }
    }
}
